pkgs/development/python-modules/pypdfium2/bin.nix

```diff
@@ -0,0 +1,152 @@
+
+  parsedNumericVersionComponents = # Turns e.g. `"121.9.6110.0"` into `[121 9 6110 0]`.
+    let jsonList = map builtins.fromJSON (lib.splitVersion pdfiumVersion);
+    in
+      assert lib.length jsonList == 4;
+      assert lib.all lib.isInt jsonList;
+      jsonList;
+  pdfiumVersion_major = lib.elemAt parsedNumericVersionComponents 0;
+  pdfiumVersion_minor = lib.elemAt parsedNumericVersionComponents 1;
+  pdfiumVersion_build = lib.elemAt parsedNumericVersionComponents 2;
+  pdfiumVersion_patch = lib.elemAt parsedNumericVersionComponents 3;
+
+  # Contains more than a single dir; `builtins.fetchTarball` and `fetchzip` cannot handle that yet,
+  # so we need to manually unpack it below.
+  pdfiumPrebuiltTar = (fetchurl {
+    url = "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium/${toString pdfiumVersion_build}/pdfium-${pdfium.platformFileInfix}.tgz";
+    hash = pdfium.hash;
+  });
+
+  pdfium-bin = runCommand "pdfium-bin" {} ''
+    mkdir "$out"
+    tar xf "${pdfiumPrebuiltTar}" --directory "$out"
+  '';
+
+  ctypesgen_pypdfium_fork_package =
+    { lib
+    , buildPythonApplication
+    , fetchFromGitHub
+    , toml
+    , setuptools-scm
+    }:
+      buildPythonApplication rec {
+        pname = "ctypesgen";
+        version = "3a811c28160b7c742f97ee50bfe55b18d4021ca0";
+
+        src = fetchFromGitHub {
+          owner = "pypdfium2-team"; # fork needed for `pypdfium2`
+          repo = "ctypesgen";
+          rev = version;
+          hash = "sha256-VqZ7Xn/7mrBjzIA7RUQt4WkrOttJMb4XmFWPpGtPfHY=";
+        };
+
+        # This package uses setuptools-scm to derive the version from the git repo (which we don't have),
+        # so use this environment variable to set it manually instead.
+        SETUPTOOLS_SCM_PRETEND_VERSION = "1.1.1+g${version}"; # we use the most recent tag and append the git version
+
+        doCheck = false;
+
+        propagatedBuildInputs = [
+          toml
+          setuptools-scm
+        ];
+      };
+  ctypesgen_pypdfium_fork = callPackage ctypesgen_pypdfium_fork_package {};
+
+in buildPythonPackage rec {
+  pname = "pypdfium2";
+  version = "4.24.0";
+
+  src = fetchFromGitHub {
+    owner = "pypdfium2-team";
+    repo = "pypdfium2";
+    rev = version;
+    hash = "sha256-DpXMCz0gh3x6Y27ciP1qZ2kYXSKwG8vjWsSQRIYUC5o=";
+    name = "pypdfium2";
+  };
+
+  # Following instructions from section
+  # > With caller-built data files (this is expected to work offline)
+  # https://github.com/pypdfium2-team/pypdfium2/tree/30c60af438b7cd90e22d42dd2ba5bffdeb568c42#install-source-caller
+  # with the addition of `--compile-libdirs`, which fixes `ctypesgen` warning:
+  #     WARNING: Could not load library 'pdfium'. Okay, I'll try to load it at runtime instead.
+  preConfigure = ''
+    "${ctypesgen_pypdfium_fork}/bin/ctypesgen" \
+      --library pdfium \
+      --compile-libdirs "${pdfium-bin}/lib" \
+      --runtime-libdirs "${pdfium-bin}/lib" \
```

Comment:
I think `--runtime-libdirs` should be `.`, supposing `${pdfium-bin}` is not available at runtime?
Or if it is, you wouldn't have to copy in the binary below.

Reply:
@mara004 It seems to work with both `--runtime-libdirs "${pdfium-bin}/lib"` and `--runtime-libdirs .`.

With `--runtime-libdirs "${pdfium-bin}/lib"`, the generated `lib/python3.11/site-packages/pypdfium2_raw/bindings.py` contains:

```python
_libdirs = ['/nix/store/q79y5h4af3vyi5kqbrwgjm6zxasshp7k-pdfium-bin/lib']
_allow_system_search = True
_libpath = _find_library("pdfium", _libdirs, _allow_system_search)
_lib = ctypes.CDLL(_libpath)
```

With `--runtime-libdirs .` the first line is

```
_libdirs = ['.']
```

instead.

> supposing `${pdfium-bin}` is not available at runtime?

The above `_libdirs = ['/nix/store/q79y5h4af3vyi5kqbrwgjm6zxasshp7k-pdfium-bin/lib']` makes `${pdfium-bin}` available at runtime.

> Or if it is, you wouldn't have to copy in the binary below.

Right -- my understanding was that `--runtime-libdirs .` with the `cp` is for bundling the `.so` inside `pypdfium2`, while `--compile-libdirs "${pdfium-bin}/lib"` would be to reference it.

Nixpkgs would likely rather reference it (better build deduplication / sharing of dependencies), but I found that removing the `cp` has the inconvenient side effect that it produces this error:

```
Version uncertain: git repo not available.
Falling back to autorelease record.
Traceback (most recent call last):
  File "/build/source/nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 154, in <module>
    main()
  File "setup.py", line 150, in main
    run_setup(modnames, pl_name, pdfium_ver)
  File "setup.py", line 134, in run_setup
    assert_exists(ModuleDir_Raw, kwargs["package_data"]["pypdfium2_raw"])
  File "setup.py", line 76, in assert_exists
    assert False, f"Missing data files: {missing}"
AssertionError: Missing data files: ['libpdfium.so']
```

This is because of:

```python
    elif pl_name == ExtPlats.system:
        kwargs["package_data"]["pypdfium2_raw"] = [VersionFN, BindingsFN]
    else:
        # ...
        kwargs["package_data"]["pypdfium2_raw"] = [VersionFN, BindingsFN, libname]
```

I'm currently setting `PDFIUM_PLATFORM='prepared!linux_x64:6110`, not `PDFIUM_PLATFORM='prepared!system:6110`, so the lower branch is taken and it expects the `libname` (`.so`) in the `pypdfium2_raw` directory.

So far from the `pypdfium2` README I understood `system` to denotate "when the system has _built_ the `.so`", and `linux_x64` to be used when using the `pdfium` binary releases, but maybe "placing the binary relase `.so` in a system directory" also means I should better use `system` instead?

Reply:
Yes, `system` is the correct one here. The latter meaning "in a system directory" is right. Explicit platforms like `linux_x64` imply bundling into pypdfium2's source tree, as the assertion hints. With `prepared!`, the target value is basically only used for the file inclusion mask.

Since I'm not used to the recipe format, I just wasn't sure about the scope of `${pdfium-bin}` (temporary to setup, or part of the install), and therefore not sure what you were trying to achieve.
But this clarifies `${pdfium-bin}` is part of the install, and you're wanting to load that at runtime, so no need to bundle an unused binary.

`--compile-libdirs` is used by ctypesgen to load the library at setup/bindings creation time to verify symbol presence, and does not impact install time library loading, which is controlled by `--runtime-libdirs` instead.
I believe you can also use `--universal-libdirs` if both match, but I haven't actually tested this yet.

And confusing though it may sound, you might want to add `--no-system-libsearch` to disable the `ctypes.util.find_library()` fallback, since you are passing the runtime libdir explicitly.
(Maybe the flag should be renamed to `--no-fallback-libsearch` to better reflect its meaning.)

Or if the library were contained in a "standard system location" as recognized by ctypes (e.g. `/usr/lib`), you could alternatively omit the explicit `--runtime-libdirs` and rely on ctypes to find it, but as NixOS library locations look quite different than on other distros, I don't think that would work, unless NixOS spoofs a different system layout (similar to conda), sets `$LD_LIBRARY_PATH` or something.

Note, if the binary is contained in a caller-given runtime libdir, pypdfium2 will of course depend on the continued presence of that dir, and load the library from there.
So if pdfium is updated and the directory's hash changes, you also have to rebuild pypdfium2 to link against the newer version of pdfium. Actually this is good because it enforces ABI safety.